src/server/transaction.cc

```diff
@@ -498,14 +494,19 @@ bool Transaction::RunInShard(EngineShard* shard, bool txq_ooo) {
   // If we're the head of tx queue (txq_ooo is false), we remove ourselves upon first invocation
   // and successive hops are run by continuation_trans_ in engine shard.
   // Otherwise we can remove ourselves only when we're concluding (so no more hops will follow).
-  bool remove_txq = is_concluding || !txq_ooo;
+  // In case of multi transaction is_concluding represents only if the current running op is
+  // concluding, therefore we remove from txq in unlock multi function which is when the transaction
+  // is concluding.
+  bool remove_txq = concluding_not_multi || !txq_ooo;
```

Comment:
The docs of RunInShard are outdated - "Returns true if the transaction should be kept in the queue." but `RunInShard` removes from the queue by itself. the same true about this function description above. A better explanation would be `returns true if the transaction continues running in the thread`.  And then probably the better name for `concluding_not_multi` is `tx_stops_running` because this is what it means, imho.